packages/frontend/workspace/src/providers/storage/index.ts

```diff
@@ -0,0 +1,29 @@
+export interface Storage {
+  /**
+   * for debug
+   */
+  name: string;
+
+  pull(
+    docId: string,
+    state: Uint8Array
+  ): Promise<{ data: Uint8Array; state?: Uint8Array } | null>;
+  push(docId: string, data: Uint8Array): Promise<void>;
+
+  /**
+   * Subscribe to updates from peer
+   *
+   * @param cb callback to handle updates
+   * @param disconnect callback to handle disconnect, reason can be something like 'network-error'
+   *
+   * @returns unsubscribe function
+   */
+  subscribe(
+    cb: (docId: string, data: Uint8Array) => void,
+    disconnect: (reason: string) => void
```

Comment:
`onDisconnected`?

Reply:
disconnect is only something you need to care after subscribe, that is the storage tell upper "we may have lost some remote updates".

Reply:
that's not what onDisconnected does?

Reply:
There does seem to be some confusion here.

For the `socketio` storage `onDisconnect` and `disconnect` callback are the same.

I consider a **stateless** storage like
`pull` and `push` are using a single http request
`subscribe` will initial a SSE(Server-Sent Events) stream transport updates
this SSE stream only exists in the current `subscribe` context and feeds back disconnect events through `disconnect` callback.

to avoid confusion,it might be better to do this?
```js
subscribe(
    cb: (event: {docId: string, data: Uint8Array} | { disconnect: true }) => void,
)
```

Reply:
never mind, I think the original signature is cool after seeing such
```ts
subscribe(
    cb: (event: {docId: string, data: Uint8Array} | { disconnect: true }) => void,
)
```